### PR TITLE
HBASE-22871 Move the DirScanPool out and do not use static field

### PR DESCRIPTION
**Pull request by Apache9**

### CODE REVIEW

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/DirScanPool.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    }
+    shutdownNow();
+    LOG.info("Update chore's pool size from {} to {}", pool.getParallelism(), size);
+    pool = new ForkJoinPool(size);
+  }
+
+  public int getSize() {
+    return size;
+  }
+}
```
Comment (Reidddddd) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/DirScanPool.java]: nit. empty line at the end of file.

**Comment by Reidddddd** [comment]
LGTM, just one nit.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 37 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 316 | master passed |
| +1 | compile | 51 | master passed |
| +1 | checkstyle | 71 | master passed |
| +1 | shadedjars | 263 | branch has no errors when building our shaded downstream artifacts. |
| +1 | javadoc | 33 | master passed |
| 0 | spotbugs | 250 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 248 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 290 | the patch passed |
| +1 | compile | 50 | the patch passed |
| +1 | javac | 50 | the patch passed |
| +1 | checkstyle | 71 | hbase-server: The patch generated 0 new + 113 unchanged - 5 fixed = 113 total (was 118) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 263 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 915 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | javadoc | 34 | the patch passed |
| +1 | findbugs | 249 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 8307 | hbase-server in the patch failed. |
| +1 | asflicense | 24 | The patch does not generate ASF License warnings. |
| | | 11602 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.replication.TestReplicationSmallTests |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/504 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 80582056ea12 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-504/out/precommit/personality/provided.sh |
| git revision | master / 43a0ec80fa |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/1/testReport/ |
| Max. process+thread count | 4687 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache9** [comment]
Any other concerns? @Reidddddd Thanks.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java [hunk 2]
```diff
@@ -1445,22 +1447,20 @@ private void startServiceThreads() throws IOException {
...
-   // Start log cleaner thread
-   int cleanerInterval = conf.getInt("hbase.master.cleaner.interval", 600 * 1000);
-   this.logCleaner =
-      new LogCleaner(cleanerInterval,
-         this, conf, getMasterWalManager().getFileSystem(),
-         getMasterWalManager().getOldLogDir());
+    // Create cleaner thread pool
+    cleanerPool = new DirScanPool(conf);
+    // Start log cleaner thread
+    int cleanerInterval = conf.getInt("hbase.master.cleaner.interval", 600 * 1000);
```
[the remark this answers is not in the document]
Comment (Apache9) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java]: I think this is just a format change? Can do it in another issue if need as it is not introduced by the patch here.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/DirScanPool.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * The thread pool used for scan directories
+ */
+@InterfaceAudience.Private
+public class DirScanPool implements ConfigurationObserver {
+  private static final Logger LOG = LoggerFactory.getLogger(DirScanPool.class);
+  private volatile int size;
+  private ForkJoinPool pool;
+  private int cleanerLatch;
+  private boolean reconfigNotification;
```
[the remark this answers is not in the document]
Comment (Apache9) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/DirScanPool.java]: It will only be accessed in the synchronzied b lock so no problem. Only size can be accessed in non-synchronized method getSize so it is volatile.

[the remark this answers is not in the document]
Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/DirScanPool.java]: OK.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/CleanerChore.java [hunk 4]
```diff
@@ -485,10 +384,13 @@ public boolean setEnabled(final boolean enabled) {
...
   /**
-   * Attemps to clean up a directory, its subdirectories, and files.
-   * Return value is true if everything was deleted. false on partial / total failures.
+   * Attemps to clean up a directory, its subdirectories, and files. Return value is true if
+   * everything was deleted. false on partial / total failures.
    */
-  private class CleanerTask extends RecursiveTask<Boolean> {
+  private final class CleanerTask extends RecursiveTask<Boolean> {
+
+    private static final long serialVersionUID = -5444212174088754172L;
```
Comment (openinx) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/CleanerChore.java]: nit: do we really need this serialVersiolUID ?  I don't see any serialization...

Comment (Apache9) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/CleanerChore.java]: The RecursiveTask implements Serializable so we need a serialVersionUID, otherwise there will be a warning.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 66 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 411 | master passed |
| +1 | compile | 70 | master passed |
| +1 | checkstyle | 94 | master passed |
| +1 | shadedjars | 308 | branch has no errors when building our shaded downstream artifacts. |
| +1 | javadoc | 42 | master passed |
| 0 | spotbugs | 301 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 300 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 345 | the patch passed |
| +1 | compile | 66 | the patch passed |
| +1 | javac | 66 | the patch passed |
| +1 | checkstyle | 100 | hbase-server: The patch generated 0 new + 113 unchanged - 5 fixed = 113 total (was 118) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 322 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 1082 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | javadoc | 38 | the patch passed |
| +1 | findbugs | 258 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 16092 | hbase-server in the patch passed. |
| +1 | asflicense | 26 | The patch does not generate ASF License warnings. |
| | | 20051 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.7 Server=18.09.7 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/504 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 18ae4df26546 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-504/out/precommit/personality/provided.sh |
| git revision | master / 43a0ec80fa |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/2/testReport/ |
| Max. process+thread count | 4675 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 64 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 407 | master passed |
| +1 | compile | 72 | master passed |
| +1 | checkstyle | 100 | master passed |
| +1 | shadedjars | 359 | branch has no errors when building our shaded downstream artifacts. |
| +1 | javadoc | 46 | master passed |
| 0 | spotbugs | 335 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 332 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 368 | the patch passed |
| +1 | compile | 68 | the patch passed |
| +1 | javac | 68 | the patch passed |
| +1 | checkstyle | 89 | hbase-server: The patch generated 0 new + 113 unchanged - 5 fixed = 113 total (was 118) |
| +1 | whitespace | 1 | The patch has no whitespace issues. |
| +1 | shadedjars | 364 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 1230 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | javadoc | 44 | the patch passed |
| +1 | findbugs | 348 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 15489 | hbase-server in the patch failed. |
| +1 | asflicense | 38 | The patch does not generate ASF License warnings. |
| | | 19946 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.replication.TestReplicationDisableInactivePeer |
|   | hadoop.hbase.replication.TestReplicationSmallTests |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/504 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 99aed17d9b1d 4.4.0-154-generic #181-Ubuntu SMP Tue Jun 25 05:29:03 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-504/out/precommit/personality/provided.sh |
| git revision | master / 7903f55c18 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/3/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/3/testReport/ |
| Max. process+thread count | 4474 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-504/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/cleaner/TestCleanerChore.java [hunk 5]
```diff
@@ -53,21 +53,22 @@
 
   @ClassRule
   public static final HBaseClassTestRule CLASS_RULE =
-      HBaseClassTestRule.forClass(TestCleanerChore.class);
```
Comment (Reidddddd) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/master/cleaner/TestCleanerChore.java]: Try avoiding this unnecessary change.
